src/session.c

```diff
@@ -104,6 +106,11 @@ static void remove_link_endpoint(LINK_ENDPOINT_HANDLE link_endpoint)
 
 static void free_link_endpoint(LINK_ENDPOINT_HANDLE link_endpoint)
 {
+    if (link_endpoint->on_link_endpoint_destroyed_callback != NULL)
+    {
```

Comment:
So, if the on_link_ep_destr_cb is set, and the link_endpoint and link (aka context) check-out (via on_link_ep_destr), then is it safe to assume that the link_endpoint has already been freed by the session?  And assume if on_link_sp_dest_cb has not been set, or link_endpoint and link do not check-out, then the link_endpoint has NOT been freed by the session? 

It's not clear to me to make this assumption (other than your comment in link.c line 712 implies I should).  Perhaps a few more comments explaining how this relationship works could help a future reader understand this.  I'm concerned a future change elsewhere in code could introduce a memory leak if the coder does not fully understand.  What do you think?

Reply:
Makes sense. I'll add the note explaining why this is needed.